Constructive Code Review/outline.txt

```diff
@@ -0,0 +1,205 @@
+- Splash screen
+    - PR screenshot: “Your code is bad and you are bad. Have a bad day.”
+- Importance
+    - Who's done code review before?
+        - Who's done it sync?
+            - We don't really do that, aside from maybe some pair
+              programming.
+        - Async?
+    - Pretty much all code at Mozilla gets reviewed before it gets
+      merged.
+        - It's a great way to…
+            - Improve the code
+                - I like to think I'm good at this, but people still
+                  find improvements to make in my stuff almost every
+                  time.
+            - Learn the codebase and the craft
+                - It informs other people what's going on in the
+                  codebase so you're not stuck with the sole
+                  responsibility of maintaining it.
+                - It's as much about teaching as fixing. It's a major
+                  way we bring junior people up the ladder.
+                - Tests may show the code works, but reviews ensure
+                  it's easy to understand.
```

Comment:
I'm not sure this is generally true. Tests reflect our understanding of the problem domain, but don't exceed those bounds. The more people that are involved in writing/editing/reviewing those tests, that increased likelihood that the tests are representative of reality. I don't think that tests by themselves show the code works.

Further, I think reviews cover a lot more than tests could ever cover. It's an opportunity *before the code lands* for someone(s) to sanity-check the mental model the person/people who wrote the code have of the problem domain in the context of the rest of what exists. A good review can point out errors in the mental model, ensure readability, ensure uniformity with other parts of the codebase (code style, architectural style, existing APIs, etc), raise awareness of similar work being done, etc. Tests can sort of cover style guidelines, but none of the rest of that.

As I look at the dichotomy in this "It's a great way to..." section, I'm not sure how my thoughts break down nicely here.